[PyTorch][Dispatcher] Fix destruction order fiasco crash

andrewjcg opened the pull request:
Summary:
The current implementation of `Dispatcher` returns an RAII object
from it's `register*` methods which, on destruction, uses a saved
reference to the `Dispatcher` to call the associated `deregister*`
method.

However, nothing guarantees that the `Dispatcher` is destroyed
*after* all RAII objects have been destroyed and, in practice, we
see segfaults caused when a global `Dispatcher` is cleaned up
before RAII globals.

This diff fixes by keeping the `Dispatcher` lock and "alive" marker
in a `std::shared_ptr` which the callbacks copy and then use to
verify that the `Dispatcher` is still alive before continuing.

https://fb.workplace.com/groups/1405155842844877/posts/7143161099044294/
https://fb.workplace.com/groups/python.builds/posts/3510588832595867/
S349108

Test Plan: CI

Differential Revision: D47113122

andrewjcg commented:
@pytorchbot label "topic: not user facing"

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -288,18 +300,24 @@ void Dispatcher::deregisterImpl_(const OperatorHandle& op, const OperatorName& o
 void Dispatcher::deregisterName_(
     const OperatorHandle& op,
     const OperatorName& op_name) {
-  std::lock_guard<std::mutex> lock(mutex_);
```
How come it's safe not to lock here anymore?

andrewjcg replied:
Lock acquisition got moved to the caller, so that the RAII instance can initiated and continue the lock before the call.

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -52,11 +52,14 @@ Dispatcher::Dispatcher()
 , listeners_(std::make_unique<detail::RegistrationListenerList>())
-, mutex_()
 , cond_var_()
+, guard_(new Guard())
```
std::make_shared and then implicit move might be more explicit?

andrewjcg committed:
[PyTorch][Dispatcher] Fix destruction order fiasco crash (#104393)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/104393

The current implementation of `Dispatcher` returns an RAII object
from it's `register*` methods which, on destruction, uses a saved
reference to the `Dispatcher` to call the associated `deregister*`
method.

However, nothing guarantees that the `Dispatcher` is destroyed
*after* all RAII objects have been destroyed and, in practice, we
see segfaults caused when a global `Dispatcher` is cleaned up
before RAII globals.

This diff fixes by keeping the `Dispatcher` lock and "alive" marker
in a `std::shared_ptr` which the callbacks copy and then use to
verify that the `Dispatcher` is still alive before continuing.

https://fb.workplace.com/groups/1405155842844877/posts/7143161099044294/
https://fb.workplace.com/groups/python.builds/posts/3510588832595867/
S349108

Test Plan: CI

Reviewed By: BrandonTheBuilder

Differential Revision: D47113122

fbshipit-source-id: 84d3be783dc6a0f3f9e2a68a7375881a2f9b64f3
